File: King Graff's Defense.cpp
/*
 MLE, 15/28, whoever set the memory limit this low is actually trolling

 https://dmoj.ca/problem/graffdefense

 Note: this is just bi-connected components, but I didn't know that when I did this problem

 - let a component be a part of the graph with no bridges inside
 - first find all bridges and remove them from the graph
 - now, any connected nodes in the graph will be part of the same component

 - there are NC2 ways to put the 2 soldiers
 - for each component, add XC2 to the valid positions, where X is the # of nodes in that component
 - answer will be (# of valid positions)/(numbers of ways to put 2 soldiers)
*/

#include<bits/stdc++.h>
using namespace std;

vector<int> adj[1000005];
int low[1000005], disc[1000005], tim;
set<pair<int, int>> bridge;

void dfs(int u, int par) {
    disc[u] = low[u] = ++tim;
    for(int v : adj[u]) {
        if(v == par) continue;
        if(!disc[v]) {
            dfs(v, u);
            low[u] = min(low[u], low[v]);
            if(low[v] > disc[u]) {
                bridge.insert({min(u, v), max(u, v)});
            }
        } else {
            low[u] = min(low[u], disc[v]);
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    cin >> n >> m;
    vector<pair<int, int>> edges;
    for(int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        if(u == v) continue;
        edges.push_back({min(u, v), max(u, v)});
        adj[u].push_back(v);
        adj[v].push_back(u);
    }
    dfs(1, -1);

    for(auto e : edges) {
        if(bridge.count(e)) {
            adj[e.first].erase(find(adj[e.first].begin(), adj[e.first].end(), e.second));
            adj[e.second].erase(find(adj[e.second].begin(), adj[e.second].end(), e.first));
        }
    }

    vector<int> comp;
    bool vis[1000005] = {0};
    for(int i = 1; i <= n; i++) {
        if(vis[i]) continue;
        int cnt = 0;
        queue<int> q;
        q.push(i);
        vis[i] = true;
        while(!q.empty()) {
            int u = q.front();
            q.pop();
            cnt++;
            for(int v : adj[u]) {
                if(vis[v]) continue;
                vis[v] = true;
                q.push(v);
            }
        }
        comp.push_back(cnt);
    }

    long long valid = 0;
    for(int c : comp) {
        valid += 1LL * c * (c - 1) / 2;
    }
    long long total = 1LL * n * (n - 1) / 2;
    cout << fixed << setprecision(5) << 1 - (double)valid / total << "\n";

    return 0;
}

File: King Graff's Defense.py
"""
MLE, 10/28, whoever set the memory limit this low is actually trolling

https://dmoj.ca/problem/graffdefense

Note: this is just bi-connected components, but I didn't know that when I did this problem

- let a component be a part of the graph with no bridges inside
- first find all bridges and remove them from the graph
- now, any connected nodes in the graph will be part of the same component

- there are NC2 ways to put the 2 soldiers
- for each component, add XC2 to the valid positions, where X is the # of nodes in that component
- answer will be (# of valid positions)/(numbers of ways to put 2 soldiers)
"""

from math import comb
import sys

input = sys.stdin.readline
sys.setrecursionlimit(500000)


def get_bridges(graph):
    n = len(graph)

    bridges = []
    low_link = [-1] * n  # smallest time reachable of a node reachable from the DFS subtree of node i
    node_id = [-1] * n  # time at which node i was discovered
    time = 0

    def dfs(cur, p):
        nonlocal time
        node_id[cur] = low_link[cur] = time
        time += 1

        for adj in graph[cur]:
            if adj == p:  # doesn't count as back edge since removing it still disconnects the graph
                continue
            if node_id[adj] == -1:  # not yet visited
                dfs(adj, cur)
                low_link[cur] = min(low_link[cur], low_link[adj])
                if low_link[adj] > node_id[cur]:  # no back edges, therefore is a bridge
                    bridges.append(frozenset((cur, adj)))

            else:  # back edge
                low_link[cur] = min(low_link[cur], node_id[adj])

    dfs(1, -1)
    return bridges


def get_components():
    vis = [False] * (N + 1)

    def dfs(start):
        cnt = 1
        stack = [start]
        vis[start] = True
        while stack:
            cur = stack.pop()
            for adj in graph[cur]:
                if vis[adj]:
                    continue
                vis[adj] = True
                cnt += 1
                stack.append(adj)
        return cnt

    comps = []
    for i in range(1, N + 1):
        if not vis[i]:
            comp_size = dfs(i)
            comps.append(comp_size)
    return comps


N, M = map(int, input().split())

edges = []
graph = [[] for _ in range(N + 1)]  # original graph
for _ in range(M):
    a, b = map(int, input().split())
    if a == b:
        continue
    edges.append(frozenset((a, b)))
    graph[a].append(b)
    graph[b].append(a)
edges = set(edges) - set(get_bridges(graph))

graph = [[] for _ in range(N + 1)]  # graph after removing bridges
for a, b in edges:
    graph[a].append(b)
    graph[b].append(a)

components = get_components()
valid = sum(comb(c, 2) for c in components)
prob = valid / comb(N, 2)
print(f"{1 - prob:.5f}")  # print to 5 decimal places with trailing zeros
